adr/0009-native-class-initialization.adoc

```diff
@@ -0,0 +1,328 @@
+
+In this scenario Quarkus core and all Quarkus extensions, including Quarkiverse extensions as well as user provided extensions along with their transitive closure of dependencies, are initialized at build time.
+This scenario reduces the entry barrier for most Quarkus users, as creating a custom Quarkus extensions is considered an advanced use case.
+Advanced users creating custom extensions are expected to be able to configure their extensions to initialize classes at build time or run time as needed.
+
+[mermaid]
+....
+---
+config:
+  look: handDrawn
+  theme: default
+---
+graph LR
+  A[Quarkus Application] --> QA[io.quarkus:quarkus-arc]
+  QA --> arc[io.quarkus.arc:arc] --> R1[...]
+  arc --> jboss[org.jboss.logging:jboss-logging]
+  QA --> core[io.quarkus:quarkus-core] --> R2[...]
+  A --> UserExtension[com.example:my-extension] --> R3[...]
+  UserExtension --> core
+  UserExtension --> jboss
+  A --> UserExtensionBT[com.example:build-time-initialized-extension] --> UserExtension
+  A --> Camel[org.apache.camel.quarkus:camel-quarkus-core]  --> R4[...]
+  A --> RestAssured[io.rest-assured:rest-assured:jar] --> R5[...]
+
+classDef white fill:#fff;
+class A white
+classDef buildtime fill:#bbf;
+class QA,arc,core,BT,Camel,R1,R2,R3,R4,UserExtension,UserExtensionBT,jboss buildtime
+....
+
+*Pros*:
+
+- Changes the current behavior as little as possible.
+- Reduces the entry barrier for most Quarkus users.
+
+*Cons*:
+
+- Keeps the entry barrier for extension developers high.
+
+==== Proposal 2: Initialize Quarkus core and a vetted list of Quarkus extensions along with their transitive dependencies at build time
+
+In this scenario, Quarkus core and core extensions, as well as a vetted list of extensions along with the transitive closure of their dependencies, are initialized at build time, while user provided extensions and third-party libraries not handled by Quarkus are initialized at runtime.
+
+[mermaid]
+....
+---
+config:
+  look: handDrawn
+  theme: default
+---
+graph LR
+  A[Quarkus Application] --> QA[io.quarkus:quarkus-arc]
+  QA --> arc[io.quarkus.arc:arc] --> R1[...]
+  arc --> jboss[org.jboss.logging:jboss-logging]
+  QA --> core[io.quarkus:quarkus-core] --> R2[...]
+  A --> UserExtension[com.example:my-extension] --> R3[...]
+  UserExtension --> core
+  UserExtension --> jboss
+  A --> UserExtensionBT[com.example:build-time-initialized-extension] --> UserExtension
+  A --> Camel[org.apache.camel.quarkus:camel-quarkus-core]  --> R4[...]
+  A --> RestAssured[io.rest-assured:rest-assured:jar] --> R5[...]
+
+classDef white fill:#fff;
+class A white
+classDef buildtime fill:#bbf;
+class QA,arc,core,BT,Camel,R1,R2,R4,UserExtensionBT,jboss buildtime
+....
+
+*Pros*:
+
+- Changes the current behavior as little as possible.
+- Reduces the entry barrier for most Quarkus users as well as extension developers.
+
+*Cons*:
+
+- Maintaining a vetted list of extensions might be cumbersome in the long run.
```

Comment:
```suggestion
- Maintaining a vetted list of extensions will be cumbersome in the long run.
```
I don't think there is a chance it won't be, especially when it's not dependencies from the core repository.